test/e2e/query_test.go

```diff
@@ -552,3 +612,55 @@ func rangeQuery(t *testing.T, ctx context.Context, addr string, q string, start,
 		return errors.Errorf("unexpected results size %d", len(res))
 	}))
 }
+
+func TestQueryMultiTenancy(t *testing.T) {
+	t.Parallel()
+
+	s, err := e2e.NewScenario("e2e_test_query_mul")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, s))
+
+	receiver1, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "1", 1)
+	receiver2, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "2", 1)
+	receiver3, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "3", 1)
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(receiver1, receiver2, receiver3))
+
+	prom1, sidecar1, err := e2ethanos.NewPrometheusWithSidecar(s.SharedDir(), s.NetworkName(), "remote-and-sidecar-1", TenentSpecificPromConfig("prom-both-remote-write-and-sidecar-1", 0, e2ethanos.RemoteWriteEndpoint(receiver1.NetworkEndpoint(8081)), "tenant-a", ""), e2ethanos.DefaultPrometheusImage())
+	testutil.Ok(t, err)
+	prom2, sidecar2, err := e2ethanos.NewPrometheusWithSidecar(s.SharedDir(), s.NetworkName(), "remote-and-sidecar-2", TenentSpecificPromConfig("prom-both-remote-write-and-sidecar-2", 0, e2ethanos.RemoteWriteEndpoint(receiver2.NetworkEndpoint(8081)), "tenant-b", ""), e2ethanos.DefaultPrometheusImage())
+	testutil.Ok(t, err)
+	prom3, sidecar3, err := e2ethanos.NewPrometheusWithSidecar(s.SharedDir(), s.NetworkName(), "remote-and-sidecar-3", TenentSpecificPromConfig("prom-both-remote-write-and-sidecar-3", 0, e2ethanos.RemoteWriteEndpoint(receiver3.NetworkEndpoint(8081)), "tenant-c", ""), e2ethanos.DefaultPrometheusImage())
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2, prom3, sidecar3))
+
+	q, err := e2ethanos.NewQuerier(s.SharedDir(), "1", []string{receiver1.GRPCNetworkEndpoint(), receiver2.GRPCNetworkEndpoint(), receiver3.GRPCNetworkEndpoint()}, nil, nil, nil, nil, nil, "", "")
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(q))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	t.Cleanup(cancel)
+
+	// Use of below function ?
+	// testutil.Ok(t, q.WaitSumMetricsWithOptions(e2e.Equals(3), []string{"thanos_store_nodes_grpc_connections"}, e2e.WaitMissingMetrics))
```

Comment:
Did this got answered?
Described here - https://github.com/thanos-io/thanos/blob/main/pkg/query/storeset.go#L151